sr_port/op_fnzpiecediff.c

```diff
@@ -0,0 +1,243 @@
+#include "mdef.h"
+#include "op.h"
+#include "stringpool.h"
+#include "matchc.h"
+#include "fnpc.h"
+
+#include "gtm_utf8.h"
+
+#define WRITE_SECTION_NUM(answer, ans_index, curr_section)			\
+{					\
+	if (ans_index != 0)	\
+	{	\
+		answer[ans_index] = ',';	\
+		ans_index++;	\
+	}	\
+	section_len = 0;	\
+	temp_section = curr_section;	\
+	while (temp_section)	\
+	{	\
+		section_len++;	\
+		temp_section /= 10;	\
+	}	\
+	temp_section_len = section_len;	\
+	temp_section = curr_section;	\
+	while (temp_section_len)	\
+	{	\
+		answer[ans_index+temp_section_len-1] = (temp_section % 10) + '0';	\
+		temp_section /= 10;	\
+		temp_section_len--;	\
+	}	\
+	ans_index += section_len;	\
+}
+
+
+#define INCREASE_ANS_LEN			\
+{					\
+	temp_section = curr_section;	\
+	section_len = 0;	\
+	while (temp_section)	\
+	{	\
+		temp_section /= 10;	\
+		section_len += 1;	\
+	}	\
+	ans_len += section_len;	\
+	if (curr_section > 1) ans_len++;	\
+}
+
+/*
+args: arg_p, arg_i, arg_len, init_matches (initial number of matches already found)
+result: delim_matches = number of delim chars matched
+*/
+#define GET_DELIM_MATCHES(arg_p, arg_i, arg_len, init_matches)		\
+{		\
+	delim_matches = init_matches;	\
+	delim_p = del->str.addr + init_matches;	\
+	temp_p = arg_p + init_matches;	\
+	temp_i = arg_i + init_matches;	\
+	while (delim_matches < delim_len && *temp_p == *delim_p && temp_i < arg_len)	\
+	{	\
+		delim_matches++;	\
+		delim_p++;	\
+		temp_p++;	\
+		temp_i++;	\
+	}	\
+	delim_p = del->str.addr;	\
+}
+
+
+void op_fnzpiecediff(mval *arg1, mval *arg2, mval *del, mval *dst) 
+{
+	const char    *arg1_p, *arg2_p, *delim_p;
+	unsigned int    arg1_len, arg2_len, delim_len;
+	unsigned int    arg1_i, arg2_i, curr_section, ans_index;
+	char    *answer;
+	
+	unsigned int    found_end;
+	unsigned int    temp_section, section_len, temp_section_len;
+	
+	unsigned int    delim_matches, temp_delim_matches;
+	unsigned int    temp_i, temp2_i;
+	const char    *temp_p, *temp2_p;
+	
+	unsigned int    *piece_diffs;
+	unsigned int    ans_len;
+	unsigned int    num_sections;
+	
+	DCL_THREADGBL_ACCESS;
+	SETUP_THREADGBL_ACCESS;
+	
+	MV_FORCE_STR(arg1);
+	MV_FORCE_STR(arg2);
+	MV_FORCE_STR(del);
+	
+	arg1_len = arg1->str.len;
+	arg2_len = arg2->str.len;
+	delim_len = del->str.len;
+	arg1_p = arg1->str.addr;
+	arg2_p = arg2->str.addr;
+	delim_p = del->str.addr;
+	arg1_i = 0;
+	arg2_i = 0;
+	curr_section = 1;
+	ans_len = 0;
+	
+	ENSURE_STP_FREE_SPACE(arg1_len*arg2_len);
+	answer = (char *)stringpool.free;
+	
+	/* allocate max possible amount of pieces */
+	if (arg1_len >= arg2_len)
+	{
+		piece_diffs = (unsigned int*)calloc(arg1_len+1, sizeof(unsigned int));
```

Comment:
Make sure to free these blocks at the end; also, please us malloc as it will be renamed to use gtm_malloc, which goes through our own layer of memory management. You will need to do a memset after.